checker/src/org/checkerframework/checker/lock/LockVisitor.java

```diff
@@ -353,9 +533,315 @@ protected boolean skipContractCheck(Tree tree, FlowExpressions.Receiver expr, Fl
         } catch (FlowExpressionParseException e) {
-            // errors are reported at declaration site
+            checker.report(e.getResult(), tree);
         }
 
         return false;
     }
+
+    @Override
+    public boolean isValidUse(AnnotatedDeclaredType declarationType,
+            AnnotatedDeclaredType useType, Tree tree) {
+        declarationType.replaceAnnotation(GUARDEDBY);
+        useType.replaceAnnotation(GUARDEDBY);
+
+        return super.isValidUse(declarationType, useType, tree);
+    }
+
+    // When visiting a method invocation, issue an error if the side effect annotation
+    // on the called method causes the side effect guarantee of the enclosing method
+    // to be violated. For example, a method annotated with @ReleasesNoLocks may not
+    // call a method annotated with @MayReleaseLocks.
+    // Also check that matching @GuardSatisfied(index) on a method's formal receiver/parameters matches
+    // those in corresponding locations on the method call site.
+    @Override
+    public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
+
+        SideEffectAnnotation seaOfInvokedMethod = atypeFactory.methodSideEffectAnnotation(TreeUtils.elementFromUse(node), false);
+
+        MethodTree enclosingMethod = TreeUtils.enclosingMethod(atypeFactory.getPath(node));
+
+        ExecutableElement methodElement = null;
+        if (enclosingMethod != null) {
+            methodElement = TreeUtils.elementFromDeclaration(enclosingMethod);
+        }
+
+        SideEffectAnnotation seaOfContainingMethod = atypeFactory.methodSideEffectAnnotation(methodElement, false);
+
+        if (atypeFactory.isWeaker(seaOfInvokedMethod, seaOfContainingMethod)) {
+            checker.report(Result.failure(
+                    "method.guarantee.violated",
+                    atypeFactory.getNameOfSideEffectAnnotation(seaOfContainingMethod),
+                    methodElement.toString(),
+                    TreeUtils.elementFromUse(node).toString(),
+                    atypeFactory.getNameOfSideEffectAnnotation(seaOfInvokedMethod)), node);
+        }
+
+        // Check that matching @GuardSatisfied(index) on a method's formal receiver/parameters matches
+        // those in corresponding locations on the method call site.
+
+        Pair<AnnotatedExecutableType, List<AnnotatedTypeMirror>> mfuPair = atypeFactory.methodFromUse(node);
+        AnnotatedExecutableType invokedMethod = mfuPair.first;
+
+        List<AnnotatedTypeMirror> requiredArgs =
+            AnnotatedTypes.expandVarArgs(atypeFactory, invokedMethod, node.getArguments());
+
+        // Index on @GuardSatisfied at each location. -1 when no @GuardSatisfied annotation was present.
+        // Note that @GuardSatisfied with no index is normally represented as having index 0.
+        // We would like to ignore it for these purposes so we convert it to -1.
+        // Unfortunately we don't distinguish between @GS and @GS(0), so if the user wrote @GS(0) we will
```

Comment:
Could you change the default value for \@GuardSatisfied to -1?  Then the user can write any non-negative integer?